src/main/java/decodes/cwms/CwmsConnectionPool.java

```diff
@@ -0,0 +1,399 @@
+
+        CwmsDbSec dbSec = CwmsDbServiceLookup.buildCwmsDb(CwmsDbSec.class, conn);
+        ResultSet rs = dbSec.getAssignedPrivGroups(conn, null);
+
+        ArrayList<StringPair> ret = new ArrayList<StringPair>();
+
+        // 4/8/13 phone call with Pete Morris - call with Null. and the columns returned are:
+        // username, user_db_office_id, db_office_id, user_group_type, user_group_owner, user_group_id,
+        // is_member, user_group_desc
+        while(rs != null && rs.next())
+        {
+            String username = rs.getString(1);
+            String db_office_id = rs.getString(2);
+            String user_group_id = rs.getString(5);
+
+            log.fine("privilegedOfficeId: username='" + username + "' "
+                + "db_office_id='" + db_office_id + "' "
+                + "user_group_id='" + user_group_id + "' "
+                );
+
+            // We look for groups "CCP Proc", "CCP Mgr", and "CCP Reviewer".
+            // Ignore anything else.
+            String gid = user_group_id.trim();
+            if (!TextUtil.startsWithIgnoreCase(gid, "CCP"))
+                continue;
+
+            // See if we have an existing privilege for this office ID.
+            int existingIdx = 0;
+            String existingPriv = null;
+            for(; existingIdx < ret.size(); existingIdx++)
+            {
+                StringPair sp = ret.get(existingIdx);
+                if (sp.first.equalsIgnoreCase(db_office_id))
+                {
+                    existingPriv = sp.second;
+                    break;
+                }
+            }
+            // If we do have an existing privilege, determine whether to keep this
+            // one or the existing one (keep the one with more privilege).
+            if (existingPriv != null)
+            {
+                if (existingPriv.toUpperCase().contains("MGR"))
+                    continue; // We are already manager in this office. Discard this item.
+                else if (gid.toUpperCase().contains("MGR"))
+                {
+                    // This item is MGR, replace existing one.
+                    ret.get(existingIdx).second = gid;
+                    continue;
+                }
+                else if (gid.toUpperCase().contains("PROC"))
+                {
+                    // This is for PROC, existing must be PROC or Reviewer. Replace.
+                    ret.get(existingIdx).second = gid;
+                    continue;
+                }
+            }
+            else // this item is first privilege seen for this office.
+                ret.add(new StringPair(db_office_id, gid));
+        }
+        return ret;
+    }
+
+    /**
+	 * Converts a CWMS String Office ID to the numeric office Code.
+	 * @param con the Connection
+	 * @param officeId the String office ID
+	 * @return the office code as a DbKey or Constants.undefinedId if no match.
+	 */
+	private DbKey officeId2code(Connection con, String officeId) throws SQLException
```

Comment:
Small efficiency here: we could cache the mapping of officeId -> code since I doubt the codes would ever change throughout the lifetime of the application.

Reply:
true, but it's also only called the once.

Reply:
Guess that's true, kinda hard to tell from the context of that single class